File: scribbleshare-commons/src/main/java/net/stzups/scribbleshare/data/objects/canvas/canvasObject/CanvasObjectType.java
package net.stzups.scribbleshare.data.objects.canvas.canvasObject;

import io.netty.util.collection.IntObjectHashMap;
import net.stzups.scribbleshare.data.exceptions.DeserializationException;
import net.stzups.scribbleshare.data.exceptions.DeserializationTypeException;

import java.util.EnumSet;
import java.util.Map;

public enum CanvasObjectType {
    SHAPE(0),
    IMAGE(1),
    MOUSE(2),
    LINE(3),
    ;

    private static final Map<Integer, CanvasObjectType> objectTypeMap = new IntObjectHashMap<>();
    static {
        for (CanvasObjectType canvasObjectType : EnumSet.allOf(CanvasObjectType.class)) {
            objectTypeMap.put(canvasObjectType.id, canvasObjectType);
        }
    }

    private final int id;

    CanvasObjectType(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    public static CanvasObjectType deserialize(int id) throws DeserializationException {
        CanvasObjectType objectType = objectTypeMap.get(id);
        if (objectType == null) {
            throw new DeserializationTypeException(CanvasObjectType.class, id);
        }
        return objectType;
    }
}

File: app/frontend/src/scripts/common/canvas/canvasObject/CanvasObjectType.ts
import ByteBuffer from "../../protocol/ByteBuffer.js";
import Shape from "./canvasObjects/Shape.js";
import CanvasImage from "./canvasObjects/CanvasImage.js";
import CanvasMouse from "./canvasObjects/CanvasMouse.js";
import LineCanvasObject from "./canvasObjects/Line.js";

enum CanvasObjectType {
    SHAPE,
    IMAGE,
    MOUSE,
    LINE,
}
export default CanvasObjectType;

export function getCanvasObject(type: CanvasObjectType, byteBuffer: ByteBuffer) {
    let object;
    switch (type) {
        case CanvasObjectType.SHAPE:
            object = new Shape(byteBuffer);
            break;
        case CanvasObjectType.IMAGE:
            object = new CanvasImage(byteBuffer);
            break;
        case CanvasObjectType.MOUSE:
            object = new CanvasMouse(byteBuffer);
            break;
        case CanvasObjectType.LINE:
            object = new LineCanvasObject(byteBuffer);
            break;
        default:
            throw new Error('unknown canvasObjectType ' + type);
    }
    return object;
}